src/hotspot/share/oops/oop.inline.hpp

```diff
@@ -277,35 +277,59 @@ bool oopDesc::is_forwarded() const {
 void oopDesc::forward_to(oop p) {
   verify_forwardee(p);
   markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversable");
+  assert(forwardee(m) == p, "encoding must be reversable");
   set_mark(m);
 }
 
-// Used by parallel scavengers
-bool oopDesc::cas_forward_to(oop p, markWord compare, atomic_memory_order order) {
-  verify_forwardee(p);
-  markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversable");
-  return cas_set_mark(m, compare, order) == compare;
+void oopDesc::forward_to_self() {
+  verify_forwardee(this);
+  markWord m = mark().set_self_forwarded();
+  assert(forwardee(m) == cast_to_oop(this), "encoding must be reversable");
+  set_mark(m);
 }
 
 oop oopDesc::forward_to_atomic(oop p, markWord compare, atomic_memory_order order) {
   verify_forwardee(p);
   markWord m = markWord::encode_pointer_as_mark(p);
-  assert(m.decode_pointer() == p, "encoding must be reversable");
+  assert(forwardee(m) == p, "encoding must be reversable");
   markWord old_mark = cas_set_mark(m, compare, order);
   if (old_mark == compare) {
     return NULL;
   } else {
-    return cast_to_oop(old_mark.decode_pointer());
+    return forwardee(old_mark);
+  }
+}
+
+oop oopDesc::forward_to_self_atomic(markWord compare, atomic_memory_order order) {
+  verify_forwardee(this);
+  while (true) {
+    markWord m = compare.set_self_forwarded();
+    assert(forwardee(m) == cast_to_oop(this), "encoding must be reversable");
+    markWord old_mark = cas_set_mark(m, compare, order);
+    if (old_mark == compare) {
+      return NULL;
+    } else if (old_mark.is_marked()) {
+      return forwardee(old_mark);
+    } else {
+      compare = old_mark;
```

Comment:
Is it even possible to get to this branch? I.e. when CAS fails, aren't we guaranteed to see `old_mark.is_marked()`?

Reply:
I don't think so. I thought a little too far here, Shenandoah-style territory. I removed this branch, and the whole loop.